src/Mvc/Mvc.Core/src/DependencyInjection/MvcCoreServiceCollectionExtensions.cs

```diff
@@ -269,6 +269,7 @@ internal static void AddMvcCoreServices(IServiceCollection services)
             //
             // Endpoint Routing / Endpoints
             //
+            services.TryAddSingleton<OrderedEndpointsSequenceProvider>();
```

Comment:
ControllerActionEndpointDataSource has an _order field to track route order. That would be unique per EndpointDataSourceBuidler.

OrderedEndpointsSequenceProvider would be shared everywhere. Have you considered what happens if there is branched middleware with multiple UseRouting/UseEndpoints?